src/command.rs

```diff
@@ -49,6 +49,42 @@ macro_rules! delegate {
     }};
 }
 
+/// If a command is `OverSsh` then it can be executed over an SSH session.
+/// Primarily a way to allow `std::process::Command` to be turned directly into an `openssh::Command`.
+pub trait OverSsh {
+    /// Given a session, return a command that can be executed over that session.
+    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session>;
+}
+
```

Comment:
I think it will be also great to have some stub implementation for `&T`, `Box<T>`, `Arc<T>` and `Rc<T>` like this:

```
impl<T: OverSsh> OverSsh for &T {
    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session> {
        <T as OverSsh>::over_session(session)
    }
}
```

Reply:
For `Box<T>`, `Arc<T>` and `Rc<T>`, we could perhaps use [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) trait, which is more extensible and will work for types defined in other crates, such as `triomphe`'s `Arc`, though I'm really not sure whether this will work, so it would need some testing:

```
impl<T: AsRef<U>, U: OverSsh> OverSsh for T {
    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session> {
        self.as_ref().over_session(session)
    }
}
```

For `&T`, we might still need to manually implement for them.

Reply:
No dice with the type checker. Seems like `U` isn't sufficiently constrained. Even this doesn't work:

```rust
impl<T: Into<U>, U: Foo> Foo for T {
    ...
}
```

I feel like I've tried to write something similar before and come across it. I'm not sure I know what its called though.

Reply:
Well, that's not unexpected...

For now let's go with manually impl for `&T`, `Box<T>`, `Arc<T>` and `Rc<T>`